File: android-twa/app/src/main/java/de/donaumoschee/app/prayer/PrayerEventId.java
package de.donaumoschee.app.prayer;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDate;
import java.util.Locale;

public final class PrayerEventId {
    private PrayerEventId() {}

    public static String create(
            String scheduleId,
            String scheduleRevision,
            LocalDate date,
            Prayer prayer,
            AlarmEvent.Kind kind,
            int leadMinutes
    ) {
        String canonical = String.join("|",
                "v2",
                scheduleId,
                scheduleRevision,
                date.toString(),
                prayer.key,
                kind.name().toLowerCase(Locale.ROOT),
                Integer.toString(leadMinutes)
        );
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256").digest(canonical.getBytes(StandardCharsets.UTF_8));
            StringBuilder hex = new StringBuilder(digest.length * 2);
            for (byte value : digest) hex.append(String.format(Locale.ROOT, "%02x", value & 0xff));
            return "p2:" + hex;
        } catch (NoSuchAlgorithmException error) {
            throw new IllegalStateException("SHA-256 unavailable", error);
        }
    }
}

File: lib/android/prayer-event-id.ts
import { createHash } from "node:crypto";

export type PrayerEventKind = "reminder" | "adhan";

export type PrayerEventIdentity = {
  scheduleId: string;
  scheduleRevision: string;
  date: string;
  prayer: string;
  kind: PrayerEventKind;
  leadMinutes: number;
  dueAtMs: number;
};

export type LegacyPrayerEventIdentity = Omit<PrayerEventIdentity, "dueAtMs">;

function sha256EventId(prefix: "p2:" | "p3:", canonical: string) {
  const digest = createHash("sha256").update(canonical, "utf8").digest("hex");
  return `${prefix}${digest}`;
}

export function prayerEventCanonicalInput(identity: PrayerEventIdentity) {
  return [
    "v3",
    identity.scheduleId,
    identity.scheduleRevision,
    identity.date,
    identity.prayer,
    identity.kind,
    String(identity.leadMinutes),
    String(identity.dueAtMs),
  ].join("|");
}

export function prayerEventId(identity: PrayerEventIdentity) {
  return sha256EventId("p3:", prayerEventCanonicalInput(identity));
}

export function legacyPrayerEventIdV2(identity: LegacyPrayerEventIdentity) {
  const canonical = [
    "v2",
    identity.scheduleId,
    identity.scheduleRevision,
    identity.date,
    identity.prayer,
    identity.kind,
    String(identity.leadMinutes),
  ].join("|");
  return sha256EventId("p2:", canonical);
}
